throughput/src/main/java/org/springframework/xd/throughput/ThroughputMessageHandler.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.xd.throughput;
+
+import static org.springframework.xd.throughput.SizeUnit.*;
+import static org.springframework.xd.throughput.TimeUnit.*;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessagingException;
+
+/**
+ * A simple handler that will count messages and log witnessed throughput at some interval.
+ *
+ * @author Eric Bottard
+ */
+public class ThroughputMessageHandler implements MessageHandler {
+
+	/*default*/ Logger logger;
+
+	private final AtomicLong counter = new AtomicLong();
+
+	private final AtomicLong start = new AtomicLong();
+
+	private final AtomicLong bytes = new AtomicLong();
+
+	private final AtomicLong intermediateStart = new AtomicLong();
+
+	private final AtomicLong intermediateCounter = new AtomicLong();
+
+	private final AtomicLong intermediateBytes = new AtomicLong();
+
+	private long reportEveryMs;
+
+	private long reportEveryNumber;
+
+	private long reportEveryBytes;
+
+	private long totalExpected;
+
+	private TimeUnit timeUnit = TimeUnit.s;
+
+	private SizeUnit sizeUnit = SizeUnit.MB;
+
+	private Clock clock = new Clock();
+
+	@Override
+	public void handleMessage(Message<?> message) throws MessagingException {
+		long now = clock.now();
+		start.compareAndSet(0L, now);
+		intermediateStart.compareAndSet(0L, now);
+		Object payload = message.getPayload();
+		long intermediateBytesP = 0L;
+		long bytesP = 0L;
+		if (payload instanceof byte[]) {
+			int size = ((byte[]) payload).length;
+			intermediateBytesP = intermediateBytes.addAndGet(size);
+			bytesP = bytes.addAndGet(size);
+		} else if (payload instanceof String) {
+			int size = ((String) payload).getBytes().length;
+			intermediateBytesP = intermediateBytes.addAndGet(size);
```

Comment:
yes, thought about that, but adds another if. Then thought that either perfectionism was the other's code-smell :)